src/crimson/net/Protocol.h

```diff
@@ -77,21 +81,58 @@ class Protocol {
     drop
   };
   void set_write_state(const write_state_t& state) {
+    if (write_state == write_state_t::open &&
+        state == write_state_t::delay) {
+      if (open_write) {
+        exit_open = seastar::shared_promise<>();
+      }
+    }
+    if (state == write_state_t::drop && exit_open) {
+      exit_open->set_value();
+      exit_open = std::nullopt;
+    }
     write_state = state;
     state_changed.set_value();
     state_changed = seastar::shared_promise<>();
   }
 
+  seastar::future<> wait_write_exit() {
+    if (exit_open) {
+      return exit_open->get_shared_future();
+    }
+    return seastar::now();
+  }
+
   void notify_keepalive_ack(utime_t keepalive_ack);
 
+  void notify_ack();
+
+  void requeue_up_to(seq_num_t seq);
+
+  void requeue_sent();
+
+  void reset_write();
+
+  bool is_queued() {
+    return conn.out_q.size() ||
```

Comment:
nit, add a parenthesis before `conn` to help editor to align the following statements to prevent other developers from reindent them by accident.

s/size()/!empty()/

Reply:
fixed, and fixed